Make/build cut-off

wclr opened the pull request:
This is an attempt to address the problem concerned in #3724. It was aslo addressed lately in #4339. With this change, the compiler will only rebuild downstream modules if they are affected by changes introduced after the previous compilation. As this is a significant build infrastructure related change I will explain the meaningful details and design decisions below.

## How build/make process currently works
- `make` gets that list of parsed modules, builds the modules dependency graph, constructs a build plan (`Language.PureScript.Make` module). 
- While constructing the build plan (`Language.PureScript.Make.BuildPlan` module), it loads existing externs files for up-to-date modules (unchanged since the previous compilation - by checking file hashes against `cahce-db`) and gathers the timestamps (previous compilation time). Based on the timestamps it determines which modules should be rebuilt. Loaded externs are used as prebuilt results in the build process.
- Then `make` compiles modules to be rebuilt.
- Waits for the result, finishes generation of new artifacts: docs, cached-db.
- Either throws the compilation errors, or returns the list of externs for all the modules (newly built and prebuilt).


## Proposed changes in the make/build process

 - While constructing the build plan, avoid loading/parsing extern files that are not needed. We just gather timestamps (compilation time) that are needed to determine what modules have to be built.

 - Then preload only needed externs: transitive dependencies of modules that need to be rebuilt, and also previously built results for modules to be rebuilt (that will be used if the module does need to be recompiled or for checking if externs have changed).

 - While module's rebuild phase check if externs of any of its dependencies have effectively changed, if not then (if the previously built result is available) we skip the compilation step.


## Avoiding loading exessive externs files
Currently externs for all the modules are loaded during construction of a build plan. Though loading/parsing those extern files may require significant (perceived) time though they are not even required for the build environment.

I noticed this problem when I added to my project a dependency with large externs files with lots of definitions (externs ~30MB in size). After just adding this package, the empty build running time (without any compilation needed) increased to more than 3 sec. With the change introduced (when those externs are not loaded) empty buuid for the described case runs for ~1 sec now.

## Changes to make API

There is a small change required (because of the above change) that adds two different make methods, that returns only updated/used externs and externs for all modules (which is needed for PSCI).

## CacheDb and illegal compiler version invalidation
Currently using info from cache-db we invalidate changed files (using timestamp/hash). Build products of incorrect compiler versions are invalidated while loading extern files. However, in the proposed changes, we also cut off the loading of not needed externs files, and we need a way to invalideate wrong compiler version products. The proposed implemented method does this by placing a compiler version in the cache-db file.

## Externs Diff
This is probably the most intricate part of the change, though essentially it is quite obvious. To know if the module has some changes, we compare externs and store the difference in so called `ExternsDiff`. When building a module we check if any of its dependencies have changed and if it uses any of those changes, if it does we rebuild it, if not we skip it. So, there are two problems to solve: 
 
- correctly determine the changes in externs - this is done by comparing structure of old and new extern's entries and considering special cases of indirect dependencies such as reexports and changes in type instances.

 - correctly check if any of the changes in dependencies affect the module - to solve this we check if the module effectively uses any of the changed elements from the imported dependencies by traversing the module's content and searching for the first occurrence of a changed element.

## Tests
To test this change comprehensively it is needed for all the elements that can be exported to check that their changes cause the rebuild for the downstream. Also check that non-effective changes do not cause the rebuild.

Another thing that needs to be tested is that while traversing the module it doesn't skip the usage of a changed element.

Existing tests have been updated to make them a bit more compact and readable.

## Status
The PR is ready for review and testing. The test spec for the change is probably quite complete, though it will need a further clean up and probably some additions considering the findings and suggestions.


---

**Checklist:**

- [ ] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

f-f commented:
I haven't looked at the code yet, but I wonder: why a new PR instead of building on top of #4339?

That code has been running in production for months now.

wclr commented:
> I haven't looked at the code yet, but I wonder: why a new PR instead of building on top of https://github.com/purescript/purescript/pull/4339?

This PR uses mostly a different approach. For example, it doesn't assume changes in externs format, etc. It provides an overview of concrete changes and explains those decisions.

wclr committed:
Fix linting error.

drathier commented:
Sorry. I've failed to explain it properly, because multiple people have looked at the code in that PR without reading the comment thread first.

I'd be more than happy for you to take a 4th stab at this, with or without the code I've written. The important thing is that we get a faster compiler.

https://github.com/drathier/purserl is where the caching fork lives, since about 6 months. It's a rewrite of the linked PR into something that's more likely to be mergable, but I've given up on getting it merged. Plus I've merged in the purerl code gen into the same binary, halfing compile times for erlang projects. We're using it daily at work and it's very useful. Merging in main every release is much less work than getting it merged, so that's what I'm doing.

wclr commented:
> Sorry. I've failed to explain it properly, because multiple people have looked at the code in that PR without reading the comment thread first.

No problem, I understand that you are using your solution in your fork, I just read that you said that code in PR itself was abandoned. I've been working on this for quite a while, even before I saw #4339. I first started working on this because I wanted to cut off loadin of unesseasary extern files during re-builds, then I saw  that I could cut off build plan itself. In #4339 I saw a solution and design decisions that were not clear to me (something that is called `caching`, and chanding externs format) so I decided to finish my work with the approach that seemed more correct to me.

> https://github.com/drathier/purserl is where the caching fork lives, since about 6 months.

You could build you version on top of this PR and compare the results.

@drathier And you also are probably the person who can competently review this PR.

drathier commented:
Yes, I'll try to find the time to review it this week.

The main idea behind the caching strategy we're using is:
1. each module exports everything that's public in the module
2. each module tracks everything it actually depended on from all imported modules when rebuilding (this is the main change to the externs file format)
3. if none of the things we're depending on changed, there's no need to recompile the current module

Step 1 and 2 are very hard to validate, as the information is lost during compilation (e.g. how do you know what type alises are you depending on when they've all been desugared away?).

https://youtu.be/BQVT6wiwCxM?t=762 is a useful high-level reference table for different caching strategies. It's also in the paper https://www.microsoft.com/en-us/research/uploads/prod/2018/03/build-systems.pdf .

wclr commented:
So if to explain the strategy I use in couple of sentences: When we have externs of a module then know all the things that dependants can use, and by comparing with previous externs results we can know what exported things have been changed/removed/added. When we decide if to recompile a module, we have all this information about its dependencies, so to determine if the module uses any of those things (exported from dependencies) that have changed.

drathier reviewed (changes requested):
I'm sure this will be better than all of the previous attempts, and I'm looking forward to getting this merged into main. The base approach seems to be the same as the one I was using, and lots of changes overlap between the two. There's lots of tests, which makes me confident that it's going to be fairly stable from day one. Main focus of this review at this stage is to make the code easier to read, so that others can understand it quickly later.

In academic software literature, a constant review speed of 200 lines an hour across all languages sure sounds slow, but here I am, spending 5 hours reviewing 1240 lines of code, some of which are renames. It keeps being spot on.

rhendric commented on src/Language/PureScript/Make.hs:
```diff
@@ -130,34 +131,64 @@ rebuildModuleWithIndex MakeActions{..} exEnv externs m@(Module _ _ moduleName _
   let toBeRebuilt = filter (BuildPlan.needsRebuild buildPlan . getModuleName . CST.resPartial) sorted
   let totalModuleCount = length toBeRebuilt
   for_ toBeRebuilt $ \m -> fork $ do
+    -- evaluate $ trace ("resPartial:" <> show (CST.resPartial $ m)) ()
+    -- evaluate $ trace ("resFull:" <> show (CST.resFull $ m)) ()
```
Nuke 'em. Commented-out code rots.

JordanMartinez replied:
This hasn't been removed. Not sure why this was marked as resolved.

JordanMartinez replied:
Ah, nevermind. I was looking at a single commit, not the entire changes made here.

rhendric commented on src/Language/PureScript/Make.hs:
```diff
@@ -183,18 +214,21 @@ make ma@MakeActions{..} ms = do
         $ M.lookup mn successes
-  return (map (lookupResult . getModuleName . CST.resPartial) sorted)
+
+  if collectAll then
+    pure $ map lookupResult sortedModuleNames
+  else
+    pure $ mapMaybe (flip M.lookup successes) sortedModuleNames
```
Move `pure` out of the `if`, though.

drathier replied:
fixed

rhendric commented on tests/TestMake.hs:
```diff
@@ -36,172 +36,510 @@ timestampB = utcMidnightOnDate 2019 1 2
+
+  let test2 fn name (mAContent1, mAContent2, mBContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          ] compile res
+
+  let testWithFailure2 fn name (mAContent1, mAContent2, mBContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          ] compileAllowingFailures res
+
+  let test3 fn name  (mAContent1, mAContent2, mBContent, mCContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          , ("C", mCContent, Nothing)
+          ] compile res
+
+  let testWithFailure3 fn name (mAContent1, mAContent2, mBContent, mCContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          , ("C", mCContent, Nothing)
+          ] compileAllowingFailures res
+
+  let recompile2 fn name ms =
+        test2 fn ("recompiles when upstream changed effectively: " <> name) ms ["A", "B"]
+
+  let recompileWithFailure2 fn name ms =
+        testWithFailure2 fn ("recompiles when upstream changed effectively: " <> name) ms ["A", "B"]
+
+  let noRecompile2 fn name ms =
+        test2 fn ("does not recompile when upstream not changed effectively: " <> name) ms ["A"]
+
   before_ (rimraf modulesDir >> rimraf sourcesDir >> createDirectory sourcesDir) $ do
     it "does not recompile if there are no changes" $ do
-      let modulePath = sourcesDir </> "Module.purs"
+      let mPath = sourcesDir </> "Module.purs"
```
This is already a large-ish PR, and I would prefer it if renames like this weren't included, at least as part of this work. (I'm not arguing that one or the other name is better; just that it's not so bad the way that it is that it needs changing.)

rhendric commented on tests/TestMake.hs:
```diff
@@ -264,8 +604,8 @@ compile input =
 compileAllowingFailures :: [FilePath] -> IO (Set P.ModuleName)
 compileAllowingFailures input = fmap snd (compileWithResult input)
 
-writeFileWithTimestamp :: FilePath -> UTCTime -> T.Text -> IO ()
-writeFileWithTimestamp path mtime contents = do
+writeFile :: FilePath -> UTCTime -> T.Text -> IO ()
+writeFile path mtime contents = do
```
Why was this renamed?

rhendric commented on src/Language/PureScript/Make.hs:
```diff
@@ -130,34 +131,64 @@ rebuildModuleWithIndex MakeActions{..} exEnv externs m@(Module _ _ moduleName _
+-- again.
+--
+-- This version will collect an return all externs of all passed modules.
+make' :: forall m. (MonadBaseControl IO m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
```
Doesn't this case correspond to the original behavior of `make`? If so, this function should be called `make` and the other one `makeAndReturnUsed` (or something); I don't think the `'` is clear enough in this case to distinguish the two behaviors, particularly if the `'` is the old behavior and the bare `make` is the new one.

I would also be fine with having `makeImp` be the new `make` and using one or the other of the constructors of `ExternsFileCollectionPolicy` (per Filip's suggestion; or whatever name you end up using) at every call site, instead of going through these two auxiliary functions.

rhendric reviewed (commented):
I want to go through this more in depth later but some quick thoughts now:

wclr commented:
❗ The thing to discuss:  currently I have added a progress message `Skipping My.ModuleB` for skipped modules (along with `Compiling My.ModuleA`). We need to decide what to do with this message, should we print it out or not.

garyb commented:
> ❗ The thing to discuss: currently I have added a progress message `Skipping My.ModuleB` for skipped modules (along with `Compiling My.ModuleA`). We need to decide what to do with this message, should we print it out or not.

I don't think messages about skipping are particularly useful outside of debugging. Compiling is useful because it's an indicator of work being done.

drathier commented:
Printing the reason why something recompiled helps people debug slow compile times. Likewise printing if compilation of a particular module took more than e.g. 10s, or printing the 10 slowest-to-compile modules. If it's easy to add something like this in, I think we should do it. Otherwise, it's a nice change in its own PR. We could add lots of debug info to the output and enable that feature using an envvar or cli flag, for example.

wclr commented:
Make method API question for discussion.

Currenlty from `Language.PureScript.Make` only one method `make` exposed that returns the list of ALL externs, that were preloaded and rebuilt.

To speed up the process, we may skip loading/parsing of significant parts of externs that are not needed for the build (preload externs only for updated modules and their transitive dependencies), which is implemented in this PR and there is an option to run make that will preload and return externs for all the modules.

Currently `make` method is used in these places:

- [app/Command/Compile.hs](https://github.com/purescript/purescript/blob/a6f6dcc05bd535937f680d48950220acbb10b5ff/app/Command/Compile.hs#L71)
Externs not used.

- [tests/TestUtils.hs](https://github.com/purescript/purescript/blob/a6f6dcc05bd535937f680d48950220acbb10b5ff/tests/TestUtils.hs#L234)
Not used either.


Here is where returned externs are used:

- [src/Language/PureScript/Interactive](https://github.com/purescript/purescript/blob/a6f6dcc05bd535937f680d48950220acbb10b5ff/src/Language/PureScript/Interactive.hs#L55)
Definitely all externs are needed.


- [tests/TestUtils.hs](https://github.com/purescript/purescript/blob/a6f6dcc05bd535937f680d48950220acbb10b5ff/tests/TestUtils.hs#L150)
Probably too.

- [src/Language/PureScript/Docs/Collect.hs](https://github.com/purescript/purescript/blob/a6f6dcc05bd535937f680d48950220acbb10b5ff/src/Language/PureScript/Docs/Collect.hs#L101)
Probably all externs required, though no tests failed when I forgot to update make method here.

The question is what would be the better API to accommodate existing cases.

Here are some options:

- Currently, I have made two methods `make` (returnes used externs) and `make'` (returns all externs).
- One `make` method with `CollectExternsPolicy` param (proposed as the first param)
- Potentially `make` method may require some other options in future, maybe consider MakeOptions record (as the first param).
- Specially named methods `makeAndReturnAll` or `makeAndReturnUsed`.

drathier commented:
Lmk when you want another review pass :) Please fix or reply to all review comments before

wclr committed:
Add comments to BuildPlan

wclr commented:
> Lmk when you want another review pass :) Please fix or reply to all review comments before

I believe I answered all the comments, not sure if I skipped some. I made updates to the code so you may review it.

JordanMartinez commented:
Hey all. Once the row error message improvements get in, this is something else I plan to look at.

drathier commented:
Sorry about the late response. Life happened.
I think we're missing these kinds of "compile, get error, fix error, recompile" regression tests? Please add :)
- https://github.com/drathier/purserl/blob/single-binary/tests/TestMake.hs#L244
- https://github.com/drathier/purserl/blob/single-binary/tests/TestMake.hs#L268
- https://github.com/drathier/purserl/blob/single-binary/tests/TestMake.hs#L316
- https://github.com/drathier/purserl/blob/single-binary/tests/TestMake.hs#L353
- https://github.com/purescript/purescript/pull/4477#discussion_r1246504879

Other than that, there's a bunch of code quality stuff, but I would be happy to adress those in another PR and get this merged and beta-shipped

wclr commented:
> Rewriting history really messes with code review change tracking, so it looks like there's 30 or so comments which are incorrectly marked as outdated. Please don't rebase/squash going forward, 

I don't remember squashing or rebasing after the comments were made, I added a couple of new commits to address the problems.

drathier commented:
To clarify, in case there was a misunderstanding. In order to merge this, I'd like at least the linked tests to be ported over, as they're actual regressions seen in that other implementation. @wclr There are many comments without replies, which I'd be happy to see a reply to too, but the regression tests are the minimum imo.

wclr commented:
Yes, I need to review the tests, as I in the comments (to comments) all the tests from your fork have passed, I will look closer at those you linked above.

> There are many comments without replies.

Not sure why many, I thought I addressed all the meaningful problems mentioned in the comments, maybe I accidentally missed some. You can probably review your comments again and resolve those that are not actual any more.

drathier commented on src/Language/PureScript/Make.hs:
```diff
@@ -130,34 +131,64 @@ rebuildModuleWithIndex MakeActions{..} exEnv externs m@(Module _ _ moduleName _
+makeImp :: forall m. (MonadBaseControl IO m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
+     => MakeActions m
+     -> [CST.PartialResult Module]
+     -> Bool
```
fixed; I can't resolve comments apparently

drathier commented on src/Language/PureScript/Make.hs:
```diff
@@ -239,7 +273,32 @@ make ma@MakeActions{..} ms = do
+                    pure $ Just (exts, MultipleErrors [], Just (emptyDiff moduleName))
+                  else
+                    pure Nothing
+                | otherwise = pure Nothing
```
all good

drathier commented on src/Language/PureScript/Make.hs:
```diff
@@ -252,8 +311,17 @@ make ma@MakeActions{..} ms = do
+            let doBuild = do
+                  (exts, warnings) <-
+                    listen $ rebuildModuleWithIndex ma env externs m (Just (idx, cnt))
+                  let diff = diffExterns exts <$> (snd <$> prevResult) <*> depsDiffs
```
unresolved question, still wondering

drathier commented on src/Language/PureScript/Make/Actions.hs:
```diff
@@ -71,16 +71,26 @@ data RebuildPolicy
+      , infx
+      , runModuleName mn
+      ]
+  SkippingModule mn mi ->
+    T.concat
+      [renderProgressIndex mi
+      , "Skipping "
+      , infx
+      , runModuleName mn
+      ]
```
unresolved, I'd like to hide this msg for now and later add a flag that enables these prints, along with explanations on why each recompile was necessary

drathier commented on src/Language/PureScript/Make/Actions.hs:
```diff
@@ -151,7 +170,7 @@ writeCacheDb'
   -- ^ The CacheDb to be written
   -> m ()
-writeCacheDb' = writeJSONFile . cacheDbFile
+writeCacheDb' = (. toCacheDbVersioned) . writeJSONFile . cacheDbFile
```
unresolved, not important

drathier commented on src/Language/PureScript/Make/Actions.hs:
```diff
@@ -234,6 +264,17 @@ buildMakeActions outputDir filePathMap foreigns usePrefix =
+    codegenTargets <- asks optionsCodegenTargets
+    let outputPaths = fmap (targetFilename mn) (S.toList codegenTargets)
+    results <- traverse (flip setTimestamp curTime) outputPaths
+    -- if something goes wrong, something failed to update, return Nothing
```
resolved

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -1,25 +1,26 @@
 
-import Control.Concurrent.Async.Lifted as A
-import Control.Concurrent.Lifted as C
+import Control.Concurrent.Async.Lifted qualified as A
+import Control.Concurrent.Lifted qualified as C
```
not important

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -88,8 +100,9 @@ markComplete
 markComplete buildPlan moduleName result = do
-  let BuildJob rVar = fromMaybe (internalError "make: markComplete no barrier") $ M.lookup moduleName (bpBuildJobs buildPlan)
-  putMVar rVar result
+  let BuildJob rVar =
+        fromMaybe (barrierError "markComplete") $ M.lookup moduleName (bpBuildJobs buildPlan)
+  C.putMVar rVar result
```
not relevant

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -103,8 +116,10 @@ collectResults
 collectResults buildPlan = do
-  let prebuiltResults = M.map (BuildJobSucceeded (MultipleErrors []) . pbExternsFile) (bpPrebuilt buildPlan)
-  barrierResults <- traverse (readMVar . bjResult) $ bpBuildJobs buildPlan
+  let mapExts exts = BuildJobSucceeded (MultipleErrors []) exts Nothing
+  let prebuiltResults =
+        M.map (mapExts . pbExternsFile) (bpPrebuilt buildPlan)
+  barrierResults <- traverse (C.readMVar . bjResult) $ bpBuildJobs buildPlan
```
not relevant

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -52,33 +54,43 @@ newtype BuildJob = BuildJob
-buildJobSuccess (BuildJobSucceeded warnings externs) = Just (warnings, externs)
+type SuccessResult = (MultipleErrors, (ExternsFile, Maybe ExternsDiff))
+
+buildJobSuccess :: BuildJobResult -> Maybe SuccessResult
+buildJobSuccess (BuildJobSucceeded warnings externs diff) = Just (warnings, (externs, diff))
 buildJobSuccess _ = Nothing
```
not important

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -113,14 +128,23 @@ getResult
+      let exts = pbExternsFile
+            $ fromMaybe (barrierError "getResult")
+            $ M.lookup moduleName (bpPrebuilt buildPlan)
+      pure (Just (MultipleErrors [], (exts, Just $ emptyDiff moduleName )))
```
improved

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -131,26 +155,75 @@ construct
   -> ([CST.PartialResult Module], [(ModuleName, [ModuleName])])
+  -> Bool
+  -- ^ If True will preload all the externs, otherwise will load only needed for
+  -- the build.
```
resolved; not an enum, but much improved by labels in wrapping options record

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -131,26 +155,75 @@ construct
+        | otherwise = M.filterWithKey (const . inBuildDeps) prebuiltMap
+
+  -- We will need previously built results for modules to be build
+  -- to skip rebuilding if deps have not changed.
```
done

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -131,26 +155,75 @@ construct
+
+  let toBeRebuilt = M.keys rebuildMap
+
+  -- Set of all dependencies of modules to be rebuilt.
+  let allBuildDeps = S.unions (S.fromList . moduleDeps <$> toBeRebuilt)
+  let inBuildDeps = flip S.member allBuildDeps
```
unresolved, question is still relevant

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -131,26 +155,75 @@ construct
+  let previous = M.mapMaybe id prevLoad
+
+  -- If for some reason (wrong version, files corruption) loading fails,
+  -- those modules should be rebuilt too.
```
done

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -160,56 +233,59 @@ construct MakeActions{..} cacheDb (sorted, graph) = do
+                _ -> toPrebuilt pb
+        where
+          toRebuild v = (M.insert mn (Just v) build, prev)
+          toPrebuilt v = (build, M.insert mn v prev)
```
not relevant

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -131,26 +155,75 @@ construct
     )
   where
+    -- Timestamp here is just to ensure that we will try to load modules that
+    -- have previous built results available.
```
not important, not fixed

drathier commented on src/Language/PureScript/Make/BuildPlan.hs:
```diff
@@ -160,56 +233,59 @@ construct MakeActions{..} cacheDb (sorted, graph) = do
+                _ | any isNothing modTimes -> toRebuild (upToDate, pb)
+                _ -> toPrebuilt pb
+        where
+          toRebuild v = (M.insert mn (Just v) build, prev)
```
done

drathier commented on src/Language/PureScript/Make/Cache.hs:
```diff
@@ -63,6 +71,35 @@ hash = ContentHash . Hash.hash
 
 type CacheDb = Map ModuleName CacheInfo
 
+data CacheDbVersioned = CacheDbVersioned { cdbVersion :: Text, cdbModules :: CacheDb }
```
not important

drathier commented on src/Language/PureScript/Make/Cache.hs:
```diff
@@ -63,6 +71,35 @@ hash = ContentHash . Hash.hash
 type CacheDb = Map ModuleName CacheInfo
 
+data CacheDbVersioned = CacheDbVersioned { cdbVersion :: Text, cdbModules :: CacheDb }
+  --deriving stock (Show)
```
done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+import Language.PureScript.Names qualified as P
+import Language.PureScript.Types qualified as P
+
+type RefStatus = Bool
```
done, enum

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+  | -- instance ref points to the class and types defined in the same module
+    -- TypeInstanceRef P.Ident (Maybe (P.ProperName 'P.ClassName)) [P.ProperName 'P.TypeName]
+    TypeInstanceRef P.Ident (ModuleName, P.ProperName 'P.ClassName) [P.ProperName 'P.TypeName]
+  deriving (Show, Eq, Ord)
```
not fixed, still relevant

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+  | -- instance ref points to the class and types defined in the same module
+    -- TypeInstanceRef P.Ident (Maybe (P.ProperName 'P.ClassName)) [P.ProperName 'P.TypeName]
+    TypeInstanceRef P.Ident (ModuleName, P.ProperName 'P.ClassName) [P.ProperName 'P.TypeName]
+  deriving (Show, Eq, Ord)
```
not relevant

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+diffExterns newExts oldExts depsDiffs =
+  ExternsDiff modName $
+    addStatus (changedRefs <> affectedReExported <> allAffectedLocalRefs)
+  where
```
not super important

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+    getRefsSet (a, r, c, u) = S.fromList $ map fst (a <> r <> c <> u)
+    fixityCtx = M.insert modName (getRefsSet declsSplit) depsDiffsMap
+
+    -- Determine which  declarations  where directly changed or removed.
```
done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+                  else case S.member clsRef <$> M.lookup clsMod depsDiffsMap of
+                    Just True ->
+                      -- if the type class is in another module and it has
+                      -- changed we don't need to care about instance types.
```
done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+        ]
+
+    removedSet = S.fromList (map fst removed)
+    changedRefs = S.fromList $ map fst (removed <> changed)
```
not important

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+    allAffectedLocalRefs =
+      S.fromList $
+        map fst $
+          filter (any (flip elem (fst <$> affectedLocalRefs)) . snd) refsGraph
```
not important, seems to be fixed

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+        map fst $
+          filter (any (flip elem (fst <$> affectedLocalRefs)) . snd) refsGraph
+
+checkDiffs :: P.Module -> [ExternsDiff] -> Bool
+checkDiffs (P.Module _ _ _ decls exports) diffs
```
not important, not fixed

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+      | otherwise -> S.singleton (mn, TypeOpRef tn)
+    P.ConstrainedType _ c _ ->
+      S.singleton (map TypeClassRef (qualified $ P.constraintClass c))
+    P.TypeOp _ _ ->
+      internalError "typeDeps: type is not qualified"
```
not important

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+      | otherwise = False
+    affectedReExported = S.fromList $ map snd $ filter checkRe oldExports
+
+    getDecls = map stripDeclaration . P.efDeclarations
```
not important, not done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+      P.ModuleRef _ mn -> not . null $ S.filter ((== Just mn) . fst) searches
+      _ -> False
+
+-- Goes though the module and try to find any usage of the refs.
+checkUsage :: Set (Maybe ModuleName, Ref) -> [P.Declaration] -> Bool
+checkUsage searches decls = foldMap findUsage decls /= mempty
```
not done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+    stripCtorType x = x
+
+    searches' = S.map (map stripCtorType) searches
+    check = (\x -> [x | x]) . flip S.member searches' . toSearched
```
not done, not sure if relevant

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+-- | Traverses imports and returns a set of refs to be searched though the
+-- module. Returns Nothing if removed refs found in imports (no need to search
+-- through the module). If an empty set is returned then no changes apply to the
+-- module.
```
not done

drathier commented on src/Language/PureScript/Make/ExternsDiff.hs:
```diff
@@ -0,0 +1,440 @@
+-- | Check if type name is a type class dictionary name.
+isDictName :: P.ProperName a -> Bool
+isDictName =
+  T.isInfixOf "$" . P.runProperName
```
code seems to be deleted from pr? so it's probably done

drathier commented:
@wclr I did a pass over my review comments to see which ones were done. I'm apparently not allowed to mark my own comments as resolved, and it felt weird deleting comments instead, so I've commented on them again pointing out if they're fixed or not.

Main thing missing is porting the regression tests linked here: https://github.com/purescript/purescript/pull/4477#issuecomment-1693500843

After that, I'd be happy to merge this, and I'm really looking forward to having much better caching in the mainline compiler!

wclr committed:
Make cut-off, the main piece

wclr committed:
Refactor externs diff and make api, fix some review sugs

i-am-the-slime commented:
@wclr Should this be reviewed again?

nwolverson commented:
I believe @MonoidMusician  had started reviewing this recently. From reading above I think there's still an outstanding point about some ported regression tests

MonoidMusician commented:
Yup I've been looking at this.

@drathier are you still unable to resolve your own comments / would you like me to mark them as resolved?

MonoidMusician reviewed (commented):
I'm still going through some of the logic, but it's looking good to me so far.

- Regarding logs: I'd be happy to have a log file in `./$OUTPUT_DIR/last_build.txt` specifically for debugging why modules are being rebuilt or skipped. That way users can have some visibility into what's going on, and it can always exist so they can debug a build retroactively and we don't have to ask them to reproduce it.
- I still need to think about how it interacts with cacheDb. Is there a particular reason to use timestamps on the externs as well? I guess better interaction with ide?
- Mostly a note to self: need to think about how this interacts with builds canceled with ctrl+c.
- From our testing, builds that fail with errors seem to rebuild everything downstream. I guess that's because those modules get deleted from the cacheDb? Would extending it with a field indicating failure help?

A few specific code comments inline.

wclr commented:
@MonoidMusician Thanks for reviewing this.

Regarding your comments/suggestions:

> - Regarding logs: I'd be happy to have a log file in ./$OUTPUT_DIR/last_build.txt specifically for debugging why modules are being rebuilt or skipped. That way users can have some visibility into what's going on, and it can always exist so they can debug a build retroactively and we don't have to ask them to reproduce it.

I'm not against such a log. To do this we need to decide on the information needed in this file. I may propose something like this:
```
Module.A compiled
- changed export: myFn # list things that changed
- changed export: MyType

Module.B compiled
- changed imported: Module.A myFn # the reason for rebuild

Module.C compiled
- changed import: Module.A MyType

Module.D skipped
```

To help further review, I will point out two main reasons why cut/off may fall short:
	
- We fail to determine things that have been changed, this may happen if something is wrong with [externs diffing logic](https://github.com/purescript/purescript/blob/2e3a49ea2ad0a55fbe05521c7b8e258dc506e66d/src/Language/PureScript/Make/ExternsDiff.hs#L202C1-L202C12)
	There we find all the things directly changed in by diffing externs entries, and then searching though all other exported things in the module, that may depend on directly changed (using graph), and we also apply some custom logic for type classes instances which is not direct.
	So, if there are no basic issues with the approach and we handle all the edge and special cases (like type class instances or type ctros) there should be no problems.
	
- We fail to [find a usage of the changed thing in the checked module](https://github.com/purescript/purescript/blob/2e3a49ea2ad0a55fbe05521c7b8e258dc506e66d/src/Language/PureScript/Make/ExternsDiff.hs#L242). There we search through the module to find encounters of any changed things found in imports. This logic is more likely to be incomplete in some way because matching is not exhaustive, so some things inside the search module could just be missed.
	
These are the points that should be thoroughly tested, though eventually it should not be difficult to have a complete set of cases, the current set should be reviewed/revised.

> - I still need to think about how it interacts with cacheDb. Is there a particular reason to use timestamps on the externs as well? I guess better interaction with ide?

Cache-db contains information on source modules, timestamps and hashes so we can check if we have already performed the build for the current module state.

Timestamps on externs are used to determine previous result compilation time, so we can check if dependencies where built after their parents or not. This may be the case, for example, when _purs-ide_ performs fast-rebuild of a changed module, and then during the build, although the hash for the module in cache-db is valid, we should invalidate all its dependencies.

> - Mostly a note to self: need to think about how this interacts with builds canceled with ctrl+c.

It will produce the results for some modules and cache-db won't be changed, the next build will fix everything, but all the modules need to be rebuilt again.

This leads to the question of necessity and convenience of a single cache-db file. I think we can eventually deprecate it in favor of per module meta files. Though we will probably need a single build result summary file for other purposes (like _ide_ checks). This is just a thought for the future.

> - From our testing, builds that fail with errors seem to rebuild everything downstream. I guess that's because those modules get deleted from the cacheDb? Would extending it with a field indicating failure help?

That's right, after the build with errors failed modules are removed from cache-db as well as their dependencies (as they are treated as skipped jobs). So the next time we run the build it will invalidate all the modules that are not in cache-db.

It is an important case that I didn't address, but I did encounter it periodically. This should be easy to fix though, we will just not remove those entries from cache-db and will treat them as successful compilations if the module is up-to-date.

wclr committed:
Added requested changes, don't recompile downstream after the error, tests updated and commented.

MonoidMusician commented:
Thanks for the updates!

That sort of information looks good for the build log.

I still need to review some of the overall logic. But the details are looking good so far.

For the declaration traversals, I found a fair bit that was missing from them. I would like to see this use the existing traversals, `accumTypes` and so on. I think the maintainability is well worth the cost of doing two traversals, since it is easier to remember to update `Traversals.hs` alone than to update `ExternsDiff.hs` too.

Here's what I'm thinking of:

<details>

```diff
diff --git a/src/Language/PureScript/Make/ExternsDiff.hs b/src/Language/PureScript/Make/ExternsDiff.hs
index 910fd1a9..165e81c5 100644
--- a/src/Language/PureScript/Make/ExternsDiff.hs
+++ b/src/Language/PureScript/Make/ExternsDiff.hs
@@ -241,21 +241,16 @@ checkDiffs (P.Module _ _ _ decls exports) diffs
 -- Goes though the module and try to find any usage of the refs.
 -- Takes a set of refs to search in module's declarations, if found returns True.
 checkUsage :: Set (Maybe ModuleName, Ref) -> [P.Declaration] -> Bool
-checkUsage searches decls = foldMap findUsage decls /= mempty
+checkUsage searches decls = anyUsages
   where
-    findUsage decl =
-      let (extr, _, _, _, _) = P.everythingWithScope goDecl goExpr goBinder mempty mempty
-       in extr mempty decl
-
-    toSearched = (,) <$> P.getQual <*> P.disqualify
-
-    -- To check data constructors we remove an origin type from it.
+    -- To check data constructors we remove an origin type from it (see `checkCtor`).
+    searches' = S.map (map stripCtorType) searches
     emptyName = P.ProperName ""
     stripCtorType (ConstructorRef _ n) = ConstructorRef emptyName n
     stripCtorType x = x
 
-    searches' = S.map (map stripCtorType) searches
-    check = Any . flip S.member searches' . toSearched
+    -- Search using the `Any` monoid (disjunction)
+    check q = Any $ S.member (P.getQual q, P.disqualify q) searches'
 
     checkType = check . map TypeRef
     checkTypeOp = check . map TypeOpRef
@@ -264,31 +259,30 @@ checkUsage searches decls = foldMap findUsage decls /= mempty
     checkCtor = check . map (ConstructorRef emptyName)
     checkClass = check . map TypeClassRef
 
-    onTypes = P.everythingOnTypes (<>) $ \case
-      P.TypeConstructor _ n -> checkType n
-      P.TypeOp _ n -> checkTypeOp n
-      P.ConstrainedType _ c _ -> checkClass (P.constraintClass c)
-      _ -> mempty
+    -- Two traversals: one to pick up usages of types, one for the rest
+    Any anyUsages =
+      foldMap checkUsageInTypes decls
+        <> foldMap checkOtherUsages decls
 
-    foldCtor f (P.DataConstructorDeclaration _ _ vars) =
-      foldMap (f . snd) vars
+    -- A nested traversal: pick up types in the module then traverse the structure of the types
+    (checkUsageInTypes, _, _, _, _) =
+      P.accumTypes $ P.everythingOnTypes (<>) $ \case
+        P.TypeConstructor _ n -> checkType n
+        P.TypeOp _ n -> checkTypeOp n
+        P.ConstrainedType _ c _ -> checkClass (P.constraintClass c)
+        _ -> mempty
 
-    constraintTypes =
-      foldMap (\c -> P.constraintArgs c <> P.constraintKindArgs c)
+    checkOtherUsages =
+      let (extr, _, _, _, _) = P.everythingWithScope goDecl goExpr goBinder mempty mempty
+      in extr mempty
 
     goDecl _ = \case
-      P.TypeDeclaration t -> onTypes (P.tydeclType t)
-      P.DataDeclaration _ _ _ _ ctors -> foldMap (foldCtor onTypes) ctors
-      P.TypeSynonymDeclaration _ _ _ t -> onTypes t
-      P.KindDeclaration _ _ _ t -> onTypes t
       P.FixityDeclaration _ (Right (P.TypeFixity _ tn _)) ->
         checkType tn
       P.FixityDeclaration _ (Left (P.ValueFixity _ (P.Qualified by val) _)) ->
         either (checkValue . P.Qualified by) (checkCtor . P.Qualified by) val
-      P.TypeClassDeclaration _ _ _ cs _ _ ->
-        foldMap onTypes (constraintTypes cs)
-      P.TypeInstanceDeclaration _ _ _ _ _ cs tc sts _ ->
-        foldMap onTypes (constraintTypes cs <> sts) <> checkClass tc
+      P.TypeInstanceDeclaration _ _ _ _ _ _ tc _ _ ->
+        checkClass tc
       _ -> mempty
 
     isLocal scope ident = P.LocalIdent ident `S.member` scope
@@ -298,7 +292,6 @@ checkUsage searches decls = foldMap findUsage decls /= mempty
         | otherwise -> checkValue n
       P.Constructor _ n -> checkCtor n
       P.Op _ n -> checkValueOp n
-      P.TypedValue _ _ t -> onTypes t
       _ -> mempty
 
     goBinder _ binder = case binder of
```

</details>

A couple tests (the second case of a type in a pattern was not handled):

```purescript
    -- Type synonym change in value.
    recompile2 it "type synonym changed in value"
      ( "module A where\ntype SynA = Int\n"
      , "module A where\ntype SynA = String\n"
      , "module B where\nimport A as A\nvalue = ([] :: Array A.SynA)\n"
      )

    -- Type synonym change in pattern.
    recompile2 it "type synonym changed in pattern"
      ( "module A where\ntype SynA = Int\n"
      , "module A where\ntype SynA = String\n"
      , "module B where\nimport A as A\nfn = \\(_ :: Array A.SynA) -> 0\n"
      )
```

For full coverage, we would need similar tests for foreign imports (values and types) and kind annotations.

wclr committed:
Update ExternsDiff.checkUsage, add tests

wclr commented:
@MonoidMusician I've added your suggestions, I agree that it would be better not to duplicate the logic if possible. Added some tests.

> For full coverage, we would need similar tests for foreign imports (values and types) and kind annotations.

I believe those couple tests didn't pass because the finding types logic was not complete, now it should be (as `accumTypes` is used). So actually, we will be testing `accumTypes`?

roryc89 commented:
I would love to get this PR merged as in our team a large chunk of every day is spent waiting for recompilation. Is there anything we can do to help get it over the line?

wclr commented:
@roryc89 Have you tried to use it for your everyday work?

roryc89 commented:
@wclr Not yet but happy to try and add any issues we find with it here.

roryc89 commented:
On first impression it works great.

i-am-the-slime commented:
@roryc89 Have you tried it?
@wclr Do you happen to have a Apple Silicon binary of this lying around?

wclr commented:
> @wclr Do you happen to have a Apple Silicon binary of this lying around?

No. Why not build on your own?

i-am-the-slime commented:
@wclr Honestly? Because compile times are fine for me and I'm too lazy to set up all the Haskell stuff.

wclr committed:
Add serialize instances for saving warnings

wclr committed:
Cut off update: warnings, build reason, log file

wclr commented:
For those who want to try it out, you may download binaries here:
https://github.com/wclr/purescript/releases

It builds in CI from https://github.com/wclr/purescript/tree/wclr/release branch.

It also publishes NPM package: `npm i @wclr/purescript` that will download binaries for you.

What to expect from running the build (e.g. `spago build`, `purs compile`):

- Compiles only changed modules with affected dependants skipping not affected downsream. In IDE/LS you may try to replace `fastRebuild` workflow for `fullBuildOnSave` to get consistent build results quicker than with the current compiler.
- Returns warnings for all modules, including those which were not re-compiled (cached in `warnings.cbor` in `output`).
- Puts build progress log into `output/compile.log`.
- Changed `cache-db.json` format.

wclr commented:
@finnhodgkin Thanks for trying this out, I'll fix the index ordering. I've also included the PR you mentioned, as promised it gives ~30% boost.

MonoidMusician committed:
Fix numbering of progress messages

MonoidMusician commented:
I've started reviewing this again. I think it is close to mergeable, with a few things that need to be fixed.

Thanks for adding `./output/compile.log`. I think having that record of the last build be always available will be helpful if users run into miscompilations.

- I made a [fix](https://github.com/MonoidMusician/purescript/commit/c5c0742e0007a65d42acfa10dea3c2d90b39ec9f) so the output numbering is roughly in order. (There are still minor race conditions, but it is a useful progress indicator again.)
- I don't understand your comment about not having to [remove failed modules](https://github.com/MonoidMusician/purescript/commit/bd8d162ff74c07955a0151e00aea9e9365103716) – without removing failed modules, the compiler outputs no information and returns exit status 0 when recompiling with no changes after a failed build.
- Possibly related: I mentioned briefly on discord that I don't see cached warnings from modules upstream of the changed files, I only see warnings from modules considered in the build plan (compiled and skipped modules now). This might be a case where users want the behavior to be configurable with a flag, maybe others should weigh in on that.

Those changes are available on [my branch](https://github.com/MonoidMusician/purescript/commits/make-cutoff/). I've been testing it with those changes, comparing the outputs between incremental builds and clean builds. So far I've seen two sources of minor diffs:

- If a file is renamed without its contents changing (e.g. spago installing a different version of the same library), its old filename can/does appear in source spans in both externs and corefn, which is undesirable. Maybe `replaceSpanName` needs to be applied to externs?
- I also see line numbers that are not updated (just in externs, not in corefn). This is not as big of a deal to me, as it might affect errors/warnings but should not affect the JavaScript output, but if you can think of a way to fix that.
- I have not tested sourcemaps or documentation yet – the stale line numbers may be a bigger issue there.

Also I ran into this error once, possibly as a result of incrementally building from output that was produced by the main compiler not this branch:
```
purs: An internal error occurred during compilation: make: getBuildReason no barrier
Please report this at https://github.com/purescript/purescript/issues
CallStack (from HasCallStack):
  error, called at src/Language/PureScript/Crash.hs:10:3 in purescript-0.15.15-8WFym9lhKoSJtoPXLBrRrY:Language.PureScript.Crash
  internalError, called at src/Language/PureScript/Make/BuildPlan.hs:108:21 in purescript-0.15.15-8WFym9lhKoSJtoPXLBrRrY:Language.PureScript.Make.BuildPlan
```

Does this case need to be an error, can it just return a build reason to force the module to rebuild?

MonoidMusician committed:
Add comments to ExternsDiff

MonoidMusician committed:
Use more sets instead of lists in ExternsDiff

thomashoneyman commented:
I had some time today to look at this see if it feels mergeable based on where @MonoidMusician left off.

In short, I think it is, and while I found a few soundness issues, I've fixed them on [`trh/make-cutoff-ready`](https://github.com/purescript/purescript/tree/trh/make-cutoff-ready), which is based on her branch and the changes below. In addition, I brought in @drathier's [regression tests](https://github.com/purescript/purescript/pull/4477#issuecomment-1693500843), added several more tests. My changes on that branch are almost all test additions, and a couple of small fixes; the core of what @wclr has built and @MonoidMusician extended I think is well-worth merging essentially as-is with these fixes.

My changes include one potentially controversial addition: a `--no-cutoff` flag which rebuilds everything downstream of a change — basically, the old compiler behavior — which can be used as an escape hatch for the first release(s) carrying this just in case something isn't right. It's a handful of lines of centralized code we can take out if we want, as I know generally `purs` doesn't carry many flags.

Here are the ported regression tests:

| purserl test | what it checks | ported link |
|---|---|---|
| `TestMake.hs#L244` | transitive type-synonym change with a local alias in the middle module | [transitively tracks type synonyms (via local synonym)](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/tests/TestMake.hs#L675) |
| `TestMake.hs#L268` | same, without a local alias (used directly) | [transitively tracks type synonyms (used directly)](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/tests/TestMake.hs#L685) |
| `TestMake.hs#L316` | instance body change (instance in the class's module) produces no downstream rebuild | [does not recompile downstream when instance implementation changed (class module)](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/tests/TestMake.hs#L766) |
| `TestMake.hs#L353` | same, diamond with the instance in the data type's module | [does not recompile downstream when instance implementation changed (type module)](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/tests/TestMake.hs#L777) |
| [#discussion_r1246504879](https://github.com/purescript/purescript/pull/4477#discussion_r1246504879) | superclass change alters dict structure should rebuild users | [tracks type class superclass changes across module boundaries](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/tests/TestMake.hs#L792) |

I also restored the cut-off cases dropped in the test refactor and added coverage for re-exports, `hiding` imports, operators, kinds/roles, and desugaring.

Next, I did a few rounds of thorough verification comparing incremental builds to clean builds and found a few soundness gaps. Each of these failed on @MonoidMusician's branch, but are fixed here.

- Re-exports through unqualified imports (`module B (module A) where import A`, i.e. how `Prelude`/`Data.*` are written) didn't match against downstream usage, so changes/removals flowing through these didn't rebuild consumers. Fixed in [`ExternsDiff.checkDiffs`/`getReExported`](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/src/Language/PureScript/Make/ExternsDiff.hs).
- Newly-added re-exports were absent from the diff, so a name newly brought into scope couldn't rebuild a module it made ambiguous.
- do/ado and unary minus desugar into `bind`/`discard`/`map`/`apply`/`pure`/`negate`, which aren't in the parsed module, so changes to those went undetected. These are now searched explicitly, qualifier-aware, in [`ExternsDiff.checkUsage`](https://github.com/purescript/purescript/blob/trh/make-cutoff-ready/src/Language/PureScript/Make/ExternsDiff.hs).

Specifically I used `registry-dev` to do a bunch of clean vs. incremental builds, which has about 742 modules across 132 packages, and then compared the `externs.cbor`, `corefn.json`, `index.js` against clean builds after performing a thorough scan editing every module and comparing the clean vs. incremental results to try and produce some stale build. That's where the gaps above were found, and after fixing them I couldn't find a way to perform an edit, build, and have an incorrectly-stale result.

I think this clears @drathier's merge bar and the review points above. Of course, happy to remove `--no-cutoff` if we don't want it.

wclr commented:
Hi, sorry for my a delayed reaction, I postponed this issue for myself as I have been busy with other projects. But as this change got even more traction now I hope we will be able to wrap it up eventually. I'm fairly interested in getting it right.

@MonoidMusician

> I don't understand your comment about not having to remove failed modules – without removing failed modules, the compiler outputs no information and returns exit status 0 when recompiling with no changes after a failed build.

I think I was acting from assumption that failed modules should and do not update their info in cache db. So they should either just be removed from new cache db or as a small optimization kept from previous cache db, then rebuild could be avoided in case of the module is fixed and its hash stays the same as a previous working version.

> I mentioned briefly on discord that I don't see cached warnings from modules upstream of the changed files, I only see warnings from modules considered in the build plan (compiled and skipped modules now). This might be a case where users want the behavior to be configurable with a flag, maybe others should weigh in on that.

Right, I didn't do load of all the warnings while make. Though it can definitely flood `build` output seems **more correct to show all the warnings for all the modules**. Not sure if it should be a subject for flag `--all-warnings` or `--only-build-warnings`. Probably front-end tools like `spago` could have options to manage this.

> If a file is renamed without its contents changing (e.g. spago installing a different version of the same library), its old filename can/does appear in source spans in both externs and corefn, which is undesirable. Maybe replaceSpanName needs to be applied to externs?

I kind of missed this. It seems strange to me why we need to have those filepaths in externs files, more over in each SPAN, I guess this is also the main contributor to bloated extern's disk size esp. for large modules. For warnings files I even cleaned them and populated while loading with actual filepath.

The problem with changed filename is essential, we have output artefacts (externs, warnings, corefn, sourcemaps, docs) coupled with module's filepath so when it is changed, even if the module it self is not required to be rebuild, we should re-generate or update those artifacts.

I see a few ways for dealing with this in the current state:

- Drop this case and treat module's filename as part of the module's code (as it is coupled with build artifacts). This not a huge deal, moving/renaming modules (without updating at least a module name) and updating libraries are not the most frequent cases, so it can be left for the future.

- Implement the update of existing artifacts. We actually have implementation for externs and warnings files. Need to cover docs (I believe docs `Module` be patched the same way using `Data` class) and source maps (update a single JSON field).

At first, I was inclined for the first option as it doesn't require additional logic and seems not a big omission, but now it seems completing the second is not difficult.

Sidenote: Ideally, I would like build artifacts be decoupled from source's filepath, though for example source maps should contain link to source file inside by standards. Some custom procedures for handing such case would be required anyway.

> I also see line numbers that are not updated (just in externs, not in corefn).

What line numbers do you mean, why should they be updated?

> Also I ran into this error once, possibly as a result of incrementally building from output that was produced by the main compiler not this branch:

Can you elaborate on this. Is there reproducible workflow?

@thomashoneyman

> My changes include one potentially controversial addition: a --no-cutoff flag 

I'm in favor of adding a flag that would run build without performing externs diffs (I proposed the same in previous posts), probably it should be named more specifically, e.g. `--no-diff-check`.

Thanks for your efforts, I will look how it is better to integrate the latest changes and tests into this PR.

wclr committed:
Fix failed module recurrent compilation.
